### PR TITLE
Cluster allocation explain API documentation

### PR DESCRIPTION
**Pull request by abeyad**
This commit updates the cluster allocation explain API documentation to
explain the new request parameters and response formats, and gives
examples of the explain API responses under various scenarios.

### CODE REVIEW

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 1]
```diff
@@ -1,203 +1,278 @@
...
 --------------------------------------------------
 
 Specify the `index` and `shard` id of the shard you would like an explanation
-for, as well as the `primary` flag to indicate whether to explain a primary or
-replica shard.
+for, as well as the `primary` flag to indicate whether to explain the primary
+shard for the given shard id or one of its replica shards.  These three request
+parameters are required.  You may also specify an optional `current_node` request
+parameter to only explain a shard that is currently located on `current_node`.
+The `current_node` can be specified as either a node id or node name.
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: Would be good to have an example using `current_node`

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 2]
```diff
@@ -1,203 +1,278 @@
...
+
+You can also have Elasticsearch explain the allocation of the first unassigned
+shard that it finds by sending an empty body for the request:
+
+[source,js]
+--------------------------------------------------
+$ curl -XGET 'http://localhost:9200/_cluster/allocation/explain'
+--------------------------------------------------
+
+If you would like to include all decisions that were factored into the final
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: I'd show an example of the output before introducing `include_yes_decisions`

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 3]
```diff
@@ -1,203 +1,278 @@
...
+--------------------------------------------------
+
+The default value for `include_yes_decisions` is `false`, which will only
+include the `no` decisions in the response.  This is generally what you would
+want, as the `no` decisions indicate why a shard is unassigned or cannot be moved,
+and including all decisions include the `yes` ones adds a lot of verbosity to the
+API's response output.
+
+Additionally, you can return information gathered by the cluster info service
+about disk usage and shard sizes by setting the `include_disk_info` parameter to
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: As above, but I'd probably introduce `include_disk_info` before `include_yes_decisions`

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 4]
```diff
@@ -1,203 +1,278 @@
...
-<10> List of node decisions factoring into final decision about the shard
+<1> The current state of the shard
+<2> The reason for the shard originally becoming unassigned
+<3> The decision to allocate the shard
+<4> The decision to allocate the shard to the particular node
+<5> The decider which led to the `no` decision for the node
+<6> An explanation as to why the decider returned a `no` decision, with a helpful hint pointing to the setting that led to the decision
 
-For a shard that is already assigned, the output looks similar to:
+The API response output for an unassigned primary shard that had a
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: ... that had previously been allocated to a node in the cluster:

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 5]
```diff
@@ -1,203 +1,278 @@
...
+  "can_allocate" : "no_valid_shard_copy",
+  "allocate_explanation" : "cannot allocate because a previous copy of the primary shard existed but can no longer be found on the nodes in the cluster"
 }
 --------------------------------------------------
-<1> Node the shard is currently assigned to
-<2> The decision is "ALREADY_ASSIGNED" because the shard is currently assigned to this node
 
-You can also have Elasticsearch explain the allocation of the first unassigned
-shard it finds by sending an empty body, such as:
+The API response output for an unassigned replica due to delayed allocation:
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: for a replica that is unassigned due to delayed allocation

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 6]
```diff
@@ -1,203 +1,278 @@
...
+          "decider" : "filter",
+          "decision" : "NO",
+          "explanation" : "node does not match index setting [index.routing.allocation.include] filters [_name:\"non_existent_node\"]"
+        }
+      ]
+    }
+  ]
+}
 --------------------------------------------------
+<1> The decision of whether the shard is allowed to remain on its current node
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: I'd just delete "The decision of "

File: docs/reference/cluster/allocation-explain.asciidoc [hunk 7]
```diff
@@ -1,203 +1,278 @@
...
+<2> The deciders that factored into the decision of why the shard is not allowed to remain on its current node
+<3> The decision of whether the shard is allowed to be allocated to another node
```
Comment (clintongormley) [inline docs/reference/cluster/allocation-explain.asciidoc]: Same as above

**Review by clintongormley — APPROVED** [review]
I'd delete the "The decision of " prefix in the callouts, plus a couple more comments.  Other than that, LGTM.

I LOVE the new explanations!

**Comment by abeyad** [comment]
Thanks @clintongormley !

I pushed a1f40055cb14c5ccd8eb697f82d2493a2570e905 to address the comments and will merge.

**Comment by abeyad** [comment]
5.x commit: dd3f5d94ee11ce25af6158adbfc000c5a342b7a9
5.2 commit: d66490eb2c88e0c3a1258ff99e83c79888212d64

**Pull request merged**